types/time.go

```diff
@@ -1965,13 +1965,15 @@ func parseSingleTimeValue(unit string, format string, strictCheck bool) (int64,
 	lf := len(format) - 1
 	// Has fraction part
 	if decimalPointPos < lf {
-		if lf-decimalPointPos >= 6 {
+		dvPre := oneToSixDigitRegex.FindString(format[decimalPointPos+1:]) // the numberical prefix of the fraction part
```

Comment:
Perhaps you don't need to use the regrex here, just check if the decimal part has `-` as prefix would be sufficient.

Reply:
But it is also useless in TiDB when facing these cases below:
```
MySQL [(none)]> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.+2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.+2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:25.980000                              |
+---------------------------------------------------------+
1 row in set (0.01 sec)

MySQL [(none)]> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.*2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.*2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:28                                     |
+---------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

MySQL [(none)]> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2./2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2./2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:28                                     |
+---------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.a2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.a2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:26                                     |
+---------------------------------------------------------+
1 row in set (0.03 sec)
```
And using the MySQL, the results are :
```
mysql> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.+2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.+2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:26                                     |
+---------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.*2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.*2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:26                                     |
+---------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2./2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2./2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:26                                     |
+---------------------------------------------------------+
1 row in set (0.00 sec)

MySQL [(none)]> select DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.a2" SECOND);
+---------------------------------------------------------+
| DATE_ADD('2007-03-28 22:08:28',INTERVAL "-2.a2" SECOND) |
+---------------------------------------------------------+
| 2007-03-28 22:08:28                                     |
+---------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

From the comparsion, we can find the rule of MySQL‘s behavior. MySQL get the numberical prefix of the fraction part. The fraction part is zero if the valid prefix is not found.

Reply:
@ichn-hu PTAL

Reply:
Cool, that make sense.